File: typescript/src/main/core/aggregator.ts
import type { ProviderPort, EventPort } from './ports';
import type { Project, Session } from './domain';
import { Ok, Err, type AsyncResult } from '../../utils/Result.js';

export class Aggregator {
  constructor(private providers: ProviderPort[], private events?: EventPort) {}

  async getProjects(): AsyncResult<Project[]> {
    try {
      const results = await Promise.all(this.providers.map(p => p.fetchProjects()));
      const failures = results.filter(r => !r.success);
      if (failures.length > 0 && failures.length === results.length) {
        return Err(failures[0].error || 'All providers failed');
      }
      const all: Project[] = [];
      for (const r of results) {
        if (r.success) all.push(...r.value);
      }
      const merged = this.mergeProjects(all);
      return Ok(merged);
    } catch (e: any) {
      return Err(e?.message || 'Aggregator error');
    }
  }

  private mergeProjects(items: Project[]): Project[] {
    const map = new Map<string, Project>();
    for (const p of items) {
      const key = `${p.provider}::${p.projectPath}`;
      const existing = map.get(key);
      if (!existing) {
        map.set(key, { ...p, sessions: [...p.sessions] });
        continue;
      }
      // Merge sessions, dedupe by (provider, sessionId)
      const seen = new Set(existing.sessions.map(s => `${s.provider}::${s.sessionId}`));
      for (const s of p.sessions) {
        const skey = `${s.provider}::${s.sessionId}`;
        if (!seen.has(skey)) {
          existing.sessions.push(s);
          seen.add(skey);
        }
      }
      // Update stats/dates conservatively
      existing.mostRecentTodoDate = max(existing.mostRecentTodoDate, p.mostRecentTodoDate);
      existing.startDate = min(existing.startDate, p.startDate);
      existing.stats = combineStats(existing, p);
      existing.pathExists = existing.pathExists || p.pathExists;
    }
    return Array.from(map.values());
  }
}

function max(a?: number, b?: number): number | undefined {
  if (a == null) return b;
  if (b == null) return a;
  return Math.max(a, b);
}
function min(a?: number, b?: number): number | undefined {
  if (a == null) return b;
  if (b == null) return a;
  return Math.min(a, b);
}
function combineStats(a: Project, b: Project) {
  const at = a.stats?.todos || 0;
  const aa = a.stats?.active || 0;
  const bt = b.stats?.todos || 0;
  const ba = b.stats?.active || 0;
  return { todos: at + bt, active: aa + ba, completed: (at + bt) - (aa + ba) };
}

File: source/dotnet/Entropic.Core/Aggregator.fs
namespace Entropic.Core

open System.Collections.Generic

/// Merges data from multiple providers, deduplicating projects
/// by (provider, projectPath) and sessions by (provider, sessionId).

// @must_test(REQ-PRV-008)
module Aggregator =

    let private optMax a b =
        match a, b with
        | None, x | x, None -> x
        | Some a', Some b' -> Some(max a' b')

    let private optMin a b =
        match a, b with
        | None, x | x, None -> x
        | Some a', Some b' -> Some(min a' b')

    let private combineStats (a: ProjectStats option) (b: ProjectStats option) : ProjectStats option =
        match a, b with
        | None, x | x, None -> x
        | Some a', Some b' ->
            let todos = a'.Todos + b'.Todos
            let active = a'.Active + b'.Active
            Some { Todos = todos; Active = active; Completed = todos - active }

    let mergeProjects (items: Project list) : Project list =
        let map = Dictionary<string, Project>()
        for p in items do
            let key = sprintf "%s::%s" p.Provider p.ProjectPath
            match map.TryGetValue(key) with
            | false, _ ->
                map.[key] <- { p with Sessions = p.Sessions }
            | true, existing ->
                let seen =
                    existing.Sessions
                    |> List.map (fun s -> sprintf "%s::%s" s.Provider s.SessionId)
                    |> Set.ofList
                let newSessions =
                    p.Sessions
                    |> List.filter (fun s ->
                        let skey = sprintf "%s::%s" s.Provider s.SessionId
                        not (Set.contains skey seen))
                let merged = {
                    existing with
                        Sessions = existing.Sessions @ newSessions
                        MostRecentTodoDate = optMax existing.MostRecentTodoDate p.MostRecentTodoDate
                        StartDate = optMin existing.StartDate p.StartDate
                        Stats = combineStats existing.Stats p.Stats
                        PathExists = Some (defaultArg existing.PathExists false || defaultArg p.PathExists false)
                }
                map.[key] <- merged
        map.Values |> Seq.toList

    let getProjects (providers: IProviderPort list) (events: IEventPort option) : Async<Result<Project list, string>> =
        async {
            let! results =
                providers
                |> List.map (fun p -> p.FetchProjects())
                |> Async.Parallel
            let results = results |> Array.toList
            let failures = results |> List.filter (fun r -> match r with Error _ -> true | Ok _ -> false)
            if failures.Length > 0 && failures.Length = results.Length then
                let firstError =
                    match failures.[0] with
                    | Error e -> e
                    | Ok _ -> "All providers failed"
                return Error firstError
            else
                let all =
                    results
                    |> List.collect (fun r -> match r with Ok projects -> projects | Error _ -> [])
                let merged = mergeProjects all
                events |> Option.iter (fun e -> try e.DataChanged() with _ -> ())
                return Ok merged
        }
